### PR TITLE
Add Post subtitle in API

### PR DESCRIPTION
**Pull request by redSpoutnik**
Add Post operation in subtitle API to allow manual upload.

**Changes**

* Add new post operation in subtitle API.
* Some of the common code has been factorized with `DownloadSubtitles`, used by the Opensubtitles plugin.

Be aware that I encountered some Issues while testing Opensubtitle plugin (in order to verify that the factorization didn't break anything):
* Opensubtitle Login doesn't seems to work since `RequestContentBytes` is no longer part of `HttpRequestOptions`.
* Subtitle search throw Error:
```
[ERR] [177] Jellyfin.Plugin.OpenSubtitles.OpenSubtitleDownloader: Invalid response type
```
Even if those errors seems unrelated, I couldn't properly test the plugin.

**Issues**

#2026 [Feature 324](https://features.jellyfin.org/posts/324/manually-add-subtitles)

### CODE REVIEW

**Comment by crobibero** [comment]
Why run `_subtitleManager.UploadSubtitle` in a background task instead of in the controller?

File: MediaBrowser.Api/Subtitles/SubtitleService.cs [hunk 1]
```diff
@@ -37,6 +37,31 @@ public class DeleteSubtitle
...
+
+        [ApiMember(Name = "Format", Description = "Format", IsRequired = true, DataType = "string", ParameterType = "path", Verb = "POST")]
+        public string Format { get; set; }
+
+        [ApiMember(Name = "IsForced", Description = "IsForced", IsRequired = true, DataType = "string", ParameterType = "path", Verb = "POST")]
+        public string IsForced { get; set; }
+
+        [ApiMember(Name = "Data", Description = "Data", IsRequired = true, DataType = "string", ParameterType = "path", Verb = "POST")]
+        public string Data { get; set; }
+
```
Comment (HavermansStef) [inline MediaBrowser.Api/Subtitles/SubtitleService.cs:62]: Unnec spacing.

File: MediaBrowser.Api/Subtitles/SubtitleService.cs [hunk 2]
```diff
@@ -37,6 +37,31 @@ public class DeleteSubtitle
...
+    [Route("/Videos/{Id}/Subtitles", "POST", Summary = "Upload an external subtitle file")]
+    [Authenticated(Roles = "admin")]
+    public class PostSubtitle : IReturnVoid
+    {
+        /// <summary>
+        /// Gets or sets the id.
+        /// </summary>
+        /// <value>The id.</value>
+        [ApiMember(Name = "Id", Description = "Item Id", IsRequired = true, DataType = "string", ParameterType = "path", Verb = "POST")]
+        public string Id { get; set; }
```
Comment (HavermansStef) [inline MediaBrowser.Api/Subtitles/SubtitleService.cs:49]: Guids are used for Ids in all other methods. Use it here for consistency.

File: MediaBrowser.Api/Subtitles/SubtitleService.cs [hunk 3]
```diff
@@ -37,6 +37,31 @@ public class DeleteSubtitle
...
+
+        [ApiMember(Name = "Language", Description = "Language", IsRequired = true, DataType = "string", ParameterType = "path", Verb = "POST")]
+        public string Language { get; set; }
+
+        [ApiMember(Name = "Format", Description = "Format", IsRequired = true, DataType = "string", ParameterType = "path", Verb = "POST")]
+        public string Format { get; set; }
+
+        [ApiMember(Name = "IsForced", Description = "IsForced", IsRequired = true, DataType = "string", ParameterType = "path", Verb = "POST")]
+        public string IsForced { get; set; }
```
Comment (HavermansStef) [inline MediaBrowser.Api/Subtitles/SubtitleService.cs:58]: Used as a bool later. Why not directly typed as a bool ?

File: MediaBrowser.Controller/Subtitles/ISubtitleManager.cs [hunk 4]
```diff
@@ -50,6 +51,11 @@ public interface ISubtitleManager
         /// </summary>
         Task DownloadSubtitles(Video video, LibraryOptions libraryOptions, string subtitleId, CancellationToken cancellationToken);
 
+        /// <summary>
+        /// Upload new subtitle.
+        /// </summary>
+        Task UploadSubtitle(Video video, SubtitleResponse response);
```
Comment (HavermansStef) [inline MediaBrowser.Controller/Subtitles/ISubtitleManager.cs:57]: Parameters not described in documentation.

File: MediaBrowser.Providers/Subtitles/SubtitleManager.cs [hunk 5]
```diff
@@ -196,6 +170,48 @@ public Task DownloadSubtitles(Video video, string subtitleId, CancellationToken
...
 
+        /// <inheritdoc />
+        public async Task UploadSubtitle(Video video, SubtitleResponse response)
+        {
+            var libraryOptions = BaseItem.LibraryManager.GetLibraryOptions(video);
+
+            await TrySaveSubtitle(video, libraryOptions, response);
+        }
+
+        private async Task TrySaveSubtitle(
```
Comment (HavermansStef) [inline MediaBrowser.Providers/Subtitles/SubtitleManager.cs:181]: INFO : Better to let task return a success-code or object. Even if not used, it's better to know whether the task was executed successfully.

**Comment by dkanada** [comment]
@redSpoutnik any chance you can fix the merge conflict?

**Comment by redSpoutnik** [comment]
@dkanada Sorry for not being active recently, I had many things to do in my personal life.
I can try to do the merge next week but I cannot promise.
Is there a new release planned soon?

**Comment by dkanada** [comment]
Nothing too soon, but I've been working through the older PRs this week trying to get everything merged.

**Review by BaronGreenback — APPROVED** [review]

File: MediaBrowser.Providers/Subtitles/SubtitleManager.cs [hunk 6]
```diff
@@ -199,6 +173,47 @@ public Task DownloadSubtitles(Video video, string subtitleId, CancellationToken
             }
         }
 
+        /// <inheritdoc />
+        public async Task UploadSubtitle(Video video, SubtitleResponse response)
```
Comment (cvium) [inline MediaBrowser.Providers/Subtitles/SubtitleManager.cs:177]:
```suggestion
        public Task UploadSubtitle(Video video, SubtitleResponse response)
```

File: MediaBrowser.Providers/Subtitles/SubtitleManager.cs [hunk 7]
```diff
@@ -199,6 +173,47 @@ public Task DownloadSubtitles(Video video, string subtitleId, CancellationToken
...
+        {
+            var libraryOptions = BaseItem.LibraryManager.GetLibraryOptions(video);
+            await TrySaveSubtitle(video, libraryOptions, response).ConfigureAwait(false);
```
Comment (cvium) [inline MediaBrowser.Providers/Subtitles/SubtitleManager.cs:180]:
```suggestion
            return TrySaveSubtitle(video, libraryOptions, response);
```

File: MediaBrowser.Providers/Subtitles/SubtitleManager.cs [hunk 8]
```diff
@@ -150,37 +150,11 @@ public Task DownloadSubtitles(Video video, string subtitleId, CancellationToken
...
-                    if (saveInMediaFolder)
-                    {
-                        savePaths.Add(Path.Combine(video.ContainingFolderPath, saveFileName));
-                    }
-
-                    savePaths.Add(Path.Combine(video.GetInternalMetadataPath(), saveFileName));
-
-                    await TrySaveToFiles(memoryStream, savePaths).ConfigureAwait(false);
-                }
+                await TrySaveSubtitle(video, libraryOptions, response);
```
Comment (cvium) [inline MediaBrowser.Providers/Subtitles/SubtitleManager.cs:157]:
```suggestion
                await TrySaveSubtitle(video, libraryOptions, response).ConfigureAwait(false);
```

**Review by cvium — APPROVED** [review]

**Pull request merged**